### PR TITLE
fix: keep current HEAD up-to-date when in bootstrap mode

### PR DESCRIPTION
**Pull request by lemmih**
## Summary of changes

Changes introduced in this pull request:

- Keep the HEAD reference updated when tipsets are validated in the bootstrap mode.
- Get rid of the file-backed set of validated blocks.

Forest has a `bootstrap` mode that is supposedly faster (or more efficient) than the regular `follow` mode. This `bootstrap` mode didn't update the HEAD reference to point to the heaviest, validated tipset, and that meant we had to keep a persistent set of validated blocks. This PR gets rid of the persistent set of validated blocks. Note that I believe the `bootstrap` mode should be deleted altogether.

## Reference issue to close (if applicable)

Closes

## Change checklist

- [x] I have performed a self-review of my own code,
- [x] I have made corresponding changes to the documentation,
- [x] I have added tests that prove my fix is effective or that my feature works (if possible),
- [x] I have made sure the [CHANGELOG][1] is up-to-date. All user-facing changes should be reflected in this document.

[1]: https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md

### CODE REVIEW

File: blockchain/chain/src/store/chain_store.rs [hunk 1]
```diff
@@ -95,8 +95,8 @@ pub struct ChainStore<DB> {
     /// File backed heaviest tipset keys
     file_backed_heaviest_tipset_keys: Mutex<FileBacked<TipsetKeys>>,
 
-    /// File backed validated blocks
-    file_backed_validated_blocks: Mutex<FileBacked<HashSet<Cid>>>,
+    /// validated blocks
+    validated_blocks: Mutex<HashSet<Cid>>,
```
[the remark this answers is not in the document]
Comment (lemmih) [reply blockchain/chain/src/store/chain_store.rs:99]: `CidHashSet` isn't 100% safe. I think we should move away from it. Or re-write it.

File: blockchain/chain_sync/src/tipset_syncer.rs [hunk 2]
```diff
@@ -1117,12 +1117,13 @@ async fn sync_messages_check_state<DB: Blockstore + Clone + Send + Sync + 'stati
...
                 &chainstore,
                 bad_block_cache,
-                full_tipset,
+                full_tipset.clone(),
                 genesis,
                 invalid_block_strategy,
             )
             .await?;
         }
+        chainstore.set_heaviest_tipset(Arc::new(full_tipset.into_tipset()))?;
```
[the remark this answers is not in the document]
Comment (lemmih) [reply blockchain/chain_sync/src/tipset_syncer.rs:1126]: Why would we want that? If validating tipset X fails, going back a few tipsets and trying again shouldn't change anything.

File: blockchain/chain/src/store/chain_store.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (lemmih) [reply blockchain/chain/src/store/chain_store.rs:99]: Hmm, actually, `CidHashSet` is fine. The risk of collisions is essentially zero.

**Review by LesnyRumcajs — APPROVED** [review]
@lemmih In what conditions did you encounter this issue? Or were you browsing the code?

**Comment by lemmih** [comment]
> @lemmih In what conditions did you encounter this issue? Or were you browsing the code?

It can be seen when you sync from a snapshot. The HEAD isn't kept up-to-date (and the set of validated blocks is only synced every 10 minutes), so quitting Forest means you lose all your progress (or the last 10 minutes).

**Review by hanabi1224 — APPROVED** [review]

**Pull request merged**